source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -49,25 +58,83 @@ Http::FilterHeadersStatus CacheFilter::decodeHeaders(Http::RequestHeaderMap& hea
   ASSERT(lookup_);
-
   ENVOY_STREAM_LOG(debug, "CacheFilter::decodeHeaders starting lookup", *decoder_callbacks_);
-  lookup_->getHeaders([this](LookupResult&& result) { onHeaders(std::move(result)); });
-  if (state_ == GetHeadersState::GetHeadersResultUnusable) {
+
+  // Used exclusively to access the RequestHeaderMap in onHeaders
+  request_headers_ = &headers;
+  getHeaders();
+
+  if (get_headers_state_ == GetHeadersState::GetHeadersResultUnusable) {
     // onHeaders has already been called, and no usable cache entry was found--continue iteration.
     return Http::FilterHeadersStatus::Continue;
   }
   // onHeaders hasn't been called yet--stop iteration to wait for it, and tell it that we stopped
   // iteration.
-  state_ = GetHeadersState::FinishedGetHeadersCall;
+  get_headers_state_ = GetHeadersState::FinishedGetHeadersCall;
   return Http::FilterHeadersStatus::StopAllIterationAndWatermark;
 }
 
 Http::FilterHeadersStatus CacheFilter::encodeHeaders(Http::ResponseHeaderMap& headers,
                                                      bool end_stream) {
-  // If lookup_ is null, the request wasn't cacheable, so the response isn't either.
-  if (lookup_ && request_allows_inserts_ && CacheabilityUtils::isCacheableResponse(headers)) {
-    // TODO(yosrym93): Add date internal header or metadata to cached responses and use it instead
-    // of the date header
+  if (skip_encoding_) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  filter_state_ = FilterState::Encoding;
+
+  // If lookup_ is null, the request wasn't cacheable, so the response isn't either
+  if (!lookup_) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  if (validating_cache_entry_ && isResponseNotModified(headers)) {
+    ASSERT(lookup_result_, "CacheFilter trying to validate a non-existent lookup result");
+
+    // Check whether the cached entry should be updated before modifying the 304 response
+    const bool should_update_cached_entry = shouldUpdateCachedEntry(headers);
+
+    // Update the 304 response status code and content-length
+    headers.setStatus(lookup_result_->headers_->getStatusValue());
+    headers.setContentLength(lookup_result_->headers_->getContentLengthValue());
+
+    // A cache entry was successfully validated; encode cached body and trailers
+    // encodeCachedResponse also adds the age header to lookup_result_
+    // so it should be called before headers are merged
+    encodeCachedResponse();
+
+    // Add any missing headers from the cached response to the 304 response
+    lookup_result_->headers_->iterate(
+        [response_headers = &headers](const Http::HeaderEntry& cached_header) {
+          // TODO(yosrym93): Try to avoid copying the header key twice
+          Http::LowerCaseString key(std::string(cached_header.key().getStringView()));
+          absl::string_view value = cached_header.value().getStringView();
+          if (!response_headers->get(key)) {
+            response_headers->setCopy(key, value);
```

Comment:
The cached header key is copied twice here, once when creating `Http::LowerCaseString key` and once inside `setCopy`. I am open to suggestions to avoid this.

The `get` function accepts a `LowerCaseString`, so I cannot pass the `cached_header.key()` directly, and there is no `setViaMove` to move the key instead of copying it.

Perhaps I am wrong in any of this, looking forward to any suggestions.

Reply:
It might be as good as it can be, depending on the limitations you have. If you have to use `setCopy`, then you have to copy into the map. Any optimization would need the single write to be in there. But, you first need to check for the (lowercased) `key` before you do that, so you're stuck.

However, what's the lifetime of `cached_header`? Or more broadly `lookup_result_->headers_`? If it outlives `response_headers`, then you're free to use `setReference()`.

(Also, although this is a wild tangent and not appropriate to address in this PR, I wonder if LowerCaseString could be changed to accept "I promise this is already lower". Since it's an attempt to never let non-lowercasing slip in, maybe that's not a good idea, but... it seems like you really could be sure that a header you previously cached would be well behaved.)

Reply:
`cached_header` is a member of a member of the CacheFilter, so it only outlives the response if the filter does. However, I think it is dangerous to rely on that.
Another though I have is to use `addViaMove` and pass to it `cached_header.key()` and `cached_header.value()` as the `lookup_result_` will not be used any more. Two problems prevent that:
- The only version of `iterate` I found is a constant version. So I cannot pass these parameters to `cached_header.key()`.
- I still have to create a `LowerCaseString` to check if the header already exists.

Reply:
This part was moved to `processSuccessfulValidation` function.